File: kafka/demo/python/misc/consumer.py
import logging
from typing import Any, Dict, List, Optional, Tuple

import conf
from confluent_kafka import (Consumer, KafkaError, KafkaException, Message,
                             TopicPartition)

from .message import MessageData

# 日志对象
logger = logging.getLogger(__name__)


def create_consumer(
    topic_name: str,
    group_id: str,
    ext_props: Optional[Dict[str, Any]] = None,
) -> Consumer:
    """
    创建消费者对象

    注意, 每次执行测试后, 都会导致 Kafka 对相关主题的分区进行再平衡操作, 这里为了方便测试,
    将 Broker 配置的 `leader.imbalance.check.interval.seconds` 设置为 `10s`,
    否则会导致每次测试启动时间较长 (等待再平衡操作)

    Args:
    - `topic_name` (`str`): 主题名称
    - `group_id` (`str`): 组 ID
    - `ext_props` (`Optional[Dict[str, Any]`], optional): 附加的消费者设置项. Defaults to `None`.

    Returns:
        Consumer: 消费者对象
    """
    props = {
        "bootstrap.servers": conf.BOOTSTRAP_SERVERS,  # 指定 Broker 地址
        "group.id": group_id,  # 指定消费者所在的消费组
        "enable.auto.commit": False,    # 关闭偏移量自动提交
        "auto.offset.reset": "earliest",  # 偏移量重置选项, 可选值包括:
                                          # smallest, earliest, beginning, largest,
                                          # latest, end, error
        "enable.partition.eof": True
    }

    if ext_props:
        # 添加扩展配置项
        props.update(ext_props)

    # 创建消费者对象
    consumer = Consumer(props)

    # 记录再平衡后分配的分区列表
    assigned_partitions: Optional[List[TopicPartition]] = None

    def on_assign(
        rebalanced_consumer: Consumer,
        partitions: List[TopicPartition],
    ) -> None:
        """
        当发生再平衡且消费者分配到指定主题的新分区时被调用, 并传入新的分区列表

        Args:
        - `rebalanced_consumer` (`Consumer`): 发生再平衡的消费者对象
        - `partitions` (`List[TopicPartition]`): 新分配的分区列表, 如果未分配新分区, 则为空
        """
        # 确认消费者对象
        assert rebalanced_consumer == consumer

        # 确认分配的分区列表不为空, 即至少分配到一个分区
        # 正常情况下应该是 3 个分区, 但不排除多次测试导致的再平衡过程中, 先分配 3 个以下的分区,
        # 在后续的再平衡继续分配新的分区
        assert len(partitions) >= 1

        # 确认分区的主题
        for p in partitions:
            assert p.topic == topic_name

        nonlocal assigned_partitions
        assigned_partitions = partitions

    def on_revoke(
        rebalanced_consumer: Consumer,
        partitions: List[TopicPartition],
    ) -> None:
        """
        当发生再平衡且消费者放弃指定主题的分区时被调用, 传入被放弃的分区列表

        Args:
        - `rebalanced_consumer` (`Consumer`): 发生再平衡的消费者对象
        - `partitions` (`List[TopicPartition]`): 主动放弃的分区列表
        """
        # 确认消费者对象
        assert rebalanced_consumer == consumer

        # 确认放弃的分区即为分配的分区
        assert partitions == assigned_partitions

        # 分区被放弃前提交一次偏移量
        consumer.commit(asynchronous=False)

    def on_lost(
        rebalanced_consumer: Consumer,
        partitions: List[TopicPartition],
    ) -> None:
        """
        会在进行意外的再平衡并且参数集合中的分区已经有新的所有者的情况下被调用

        Args:
        - `rebalanced_consumer` (`Consumer`): 发生再平衡的消费者对象
        - `partitions` (`List[TopicPartition]`): 丢失的分区列表
        """
        assert rebalanced_consumer == consumer

    # 指定消费者订阅主题
    consumer.subscribe(
        [topic_name],  # 要定义的主题列表
        on_assign=on_assign,  # 再平衡后分配新分区的回调
        on_revoke=on_revoke,  # 再平衡后失去分区的回调
        on_lost=on_lost,  # 意外导致丢失分区的回调
    )

    return consumer


def poll_and_assert(  # NOSONAR
    consumer: Consumer,
    expected_messages: Optional[List[MessageData]] = None,
    commit=True,
) -> List[Message]:
    """
    从消费者读取消息并对消息进行断言

    Args:
    - `consumer` (`Consumer`): 消费者对象
    - `msg_data` (`Optional[MessageData]`, optional): 消息元组. Defaults to `None`.
    - `commit` (`bool`, optional): 是否提交偏移量. Defaults to `True`.

    Returns:
    - Message: 消息对象
    """

    # 最后读取的消息对象
    messages: List[Message] = []

    if expected_messages is None:
        expected_messages = []

    # 记录已经到达末尾的分区字典
    # Key 为主题 + 分区, Value 为 Boolean 类型
    eof: Dict[Tuple[str, int], bool] = {}

    # 读取消息, 直到将所分配分区的消息读完
    while 1:
        # 读取消息, 超时时间 2s
        msg = consumer.poll(timeout=2.0)

        if msg:
            # 获取消息的主题, 分区和错误信息
            topic, partition, err = msg.topic(), msg.partition(), msg.error()
            if err:
                # 如果已经到达分区结尾, 则退出循环, 否则抛出异常
                if err.code() != KafkaError._PARTITION_EOF:
                    raise KafkaException(err)

                logger.error(err)

                # 记录到达末尾的主题和分区
                eof[(topic, partition)] = True

                # 判断如果所有主题分区都到达末尾, 则结束循环
                if len(eof) == len(consumer.assignment()):
                    break

                continue

            # 收到消息, 将指定主题分区的 EOF 标记取消
            eof.pop((topic, partition), None)

            if expected_messages and len(messages) == len(expected_messages):
                messages = messages[1:]

            # 记录最后一条消息
            messages.append(msg)

    assert messages

    if commit:
        try:
            # 提交偏移量 (同步方式)
            consumer.commit(asynchronous=False)
        except KafkaException as e:
            if e.args[0].code() == KafkaError._NO_OFFSET:
                logger.error(e)
            else:
                raise e

    # 获取最后一条消息
    last_msg = messages[-1]

    # 确认获取的消息无错误
    assert last_msg.error() is None

    # 确认获取消息的 Key 和 Value
    if expected_messages:
        for m, em in zip(messages, expected_messages):
            assert m.key() == em.key
            assert m.value() == em.value

    return messages


def close_consumer(consumer: Consumer) -> None:
    """
    关闭消费者对象

    Args:
    - `consumer` (`Consumer`): 消费者对象

    Raises:
    - `e`: 关闭时发生的异常
    """
    try:
        consumer.close()
    except KafkaException as e:
        logger.error(e)
        if e.args[0].code() != KafkaError._NO_OFFSET:
            raise e

File: kafka/demo/go/misc/consumer.go
package misc

import (
	"context"
	"fmt"

	"github.com/confluentinc/confluent-kafka-go/kafka"
)

// # 创建消费者对象
//
// 本函数用于创建消费者对象, 并订阅指定的主题
//
// 参数:
//   - `topic`: 要订阅的主题名称
//   - `extConf`: 扩展配置项, 应包含除 `"bootstrap.servers"` 外的其它消费者配置项
//
// 返回消费者对象以及错误对象
func CreateConsumer(topic string, extConf *kafka.ConfigMap) (*kafka.Consumer, error) {
	// 创建消费者对象
	consumer, err := kafka.NewConsumer(NewConfig(extConf))
	if err != nil {
		return nil, err
	}

	// 订阅给定主题
	err = consumer.Subscribe(topic, nil)
	if err != nil {
		return nil, err
	}

	return consumer, nil
}

// # 通过指定生产者发送指定消费者的当前偏移量
//
// 在指定生产者启动事务的前提下, 事务中的消费者不能直接提交偏移量, 而应该将偏移量作为事务的一部分, 由生产者提交给集群管理器
//
// 参数:
//   - `ctx`: 异步上下文对象, 可以设置超时时间, 或者为 `context.TODO()` 设置无限等待
//   - `producer`: 启动事务的生产者对象
//   - `consumer`: 要提交偏移量的消费者对象
//
// 返回错误对象
func SendConsumerOffsetsToTransaction(ctx context.Context, producer *kafka.Producer, consumer *kafka.Consumer) error {
	// 获取消费者的所有主题分区
	tp, err := consumer.Assignment()
	if err != nil {
		return err
	}

	// 获取消费者所有主题分区的当前偏移量
	tp, err = consumer.Position(tp)
	if err != nil {
		return err
	}

	// 获取 input 消费者的群组元数据
	meta, err := consumer.GetConsumerGroupMetadata()
	if err != nil {
		return err
	}

	// 通过 output 生产者将 input 消费者的偏移量发送到 Kafka 群组
	return producer.SendOffsetsToTransaction(ctx, tp, meta)
}

// # 读取指定主题的最后 `N` 条消息
//
// 根据给定的消费者对象和消息条数 `N`, 读取该消费者订阅主题的最后 N 条消息, 返回实际数量 (`<=N`) 的消息集合
//
// 参数:
//   - `consumer`: 消费者对象
//   - `n`: 要获取的最新消息的条数
//   - `commit`: 接收完消息后是否提交偏移量
//
// 返回消息对象集合即错误对象
func PollLastNMessage(consumer *kafka.Consumer, n int, commit bool) ([]*kafka.Message, error) {
	// 创建消息切片集合
	msgs := make([]*kafka.Message, 0)

	// 用于存放到达末尾的主题分区的字典对象, Key 为主题分区组合
	eofs := make(map[string]bool)

	// 循环, 不断获取消息对象
	for stop := false; !stop; {
		// 获取一条消息, 返回结果
		ev := consumer.Poll(5000)

		// 根据结果的类型进行处理
		switch e := ev.(type) {
		case *kafka.Message: // 消息结果为 kafka.Message 类型, 表示成功获取一条消息
			// 判断消息的主题分区是否正确
			if e.TopicPartition.Error != nil {
				return nil, e.TopicPartition.Error
			}
			// 保证返回的消息数量为 N
			if len(msgs) == n {
				msgs = msgs[1:]
			}

			// 将本次取得的消息进行保存
			msgs = append(msgs, e)

			// 从 EOFs 字典中删除有消息的主题分区
			delete(eofs, fmt.Sprintf("%v|%v", *e.TopicPartition.Topic, e.TopicPartition.Partition))
		case error: // 消息结果为 error 类型, 表示发生错误
			return nil, e
		case kafka.PartitionEOF: // 消息结果为 kafka.PartitionEOF 类型, 表示一个分区到达结尾
			// 获取消费者订阅的主题和分区集合
			tps, err := consumer.Assignment()
			if err != nil {
				return nil, err
			}

			// 将以达到结尾的主题和分区存入 EOFs 字典中
			eofs[fmt.Sprintf("%v|%v", *e.Topic, e.Partition)] = true

			// 如果所有分区都到达结尾, 则退出循环
			if len(eofs) == len(tps) {
				stop = true
			}
		}
	}

	// 提交偏移量
	if commit {
		_, err := consumer.Commit()
		if err != nil {
			return nil, err
		}
	}

	return msgs, nil
}
